google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java

```diff
@@ -0,0 +1,484 @@
+  }
+
+  @Before
+  public void before() {
+    String autoId = LocalFirestoreHelper.autoId();
+    String collPath = String.format("java-%s-%s", testName.getMethodName(), autoId);
+    randomColl = firestore.collection(collPath);
+  }
+
+  @AfterClass
+  public static void afterClass() throws Exception {
+    firestore.close();
+  }
+
+  /*
+  Start a listener on a query with an empty result set
+   */
+  @Test
+  public void emptyResults() throws InterruptedException {
+    final Query query = randomColl.whereEqualTo("foo", "bar");
+    // register the snapshot listener for the query
+    QuerySnapshotEventListener listener = new QuerySnapshotEventListener(1);
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
+      listener.eventsCDL.await();
+    } finally {
+      registration.remove();
+    }
+
+    assertThat(listener.receivedEvents).hasSize(1);
+    ListenerEvent e = listener.receivedEvents.get(0);
+    assertThat(e.error).isNull();
+    QuerySnapshot snap = e.value;
+    assertThat(snap).isNotNull();
+    assertThat(snap).isEmpty();
+    assertThat(snap.getDocumentChanges()).isEmpty();
+  }
+
+  /*
+  Start a listener on a query with a non-empty result set
+   */
+  @Test
+  public void nonEmptyResults() throws InterruptedException, TimeoutException, ExecutionException {
+    // create a document in our collection that will match the query
+    randomColl.document("doc").set(map("foo", "bar")).get(5, TimeUnit.SECONDS);
+
+    // run our test
+    final Query query = randomColl.whereEqualTo("foo", "bar");
+    QuerySnapshotEventListener listener = new QuerySnapshotEventListener(1);
+    List<ListenerEvent> receivedEvents = listener.receivedEvents;
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
+      listener.eventsCDL.await();
+    } finally {
+      registration.remove();
+    }
+
+    assertThat(receivedEvents).hasSize(1);
+    ListenerEvent e = receivedEvents.get(0);
+    assertThat(e.error).isNull();
+    QuerySnapshot snap = e.value;
+    assertThat(snap).isNotNull();
+    assertThat(snap).hasSize(1);
+  }
+
+  /*
+  Starting from a query with empty results creating a new document
+  that matches the query should result in an ADDED event
+   */
+  @Test
+  public void emptyResults_newDocument_ADDED()
+      throws InterruptedException, TimeoutException, ExecutionException {
+
+    // run our test
+    final Query query = randomColl.whereEqualTo("foo", "bar");
+    QuerySnapshotEventListener listener = new QuerySnapshotEventListener(1, 1, 0, 0);
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
+      randomColl.document("doc").set(map("foo", "bar")).get(5, TimeUnit.SECONDS);
+      listener.eventsCDL.await(DocumentChange.Type.ADDED);
+    } finally {
+      registration.remove();
+    }
+
+    ListenerAssertions la = listener.assertions();
+    la.noError();
+    la.eventCountIsAnyOf(Range.closed(1, 2));
```

Comment:
If you wait for the initial empty event, the number of events is deterministic.

Reply:
Will update